src/adr/0018-polyglot-strategy.md

```diff
@@ -0,0 +1,158 @@
+# 16. Polyglot codebase
+
+Date: 2020-02-03
+
+## Status
+
+Proposed
+
+## Context & Problem Statement
+
+James is written in Java for a very long time. In recent years, Java modernized a lot after a decade of slow progress.
+
+However, in the meantime, most software relying on the JVM started supporting alternative JVM languages to keep being relevant.
+
+It includes Groovy, Clojure, Scala and more recently Kotlin.
+
+Not being open to those alternative languages can be a problem for James adoption.
+
+## Decision drivers
+
+Nowadays, libraries and framework targeting the JVM are expected to support usage of one or several of these alternative languages.
+
+James being not only a mail server but also a development framework needs to reach those expectations.
+
+At the same time, more and more developers and languages adopt Function Programming (FP) idioms to solve their problems.
+
+## Considered options
+
+1. Let the users figure out how to make polyglot setups
+2. Document the usage of polyglot mailets for some popular languages
+3. Document the usage of polyglot components for some popular languages
+4. Actually implement some mailets in some popular languages
+5. Actually implement some components in some popular languages
+
+About Languages:
+
+I. Clojure
+II. Groovy
+III. Kotlin
+IV. Scala
+
+## Decision
+
+We decided on options 4, 5 and IV.
+
+That means we need to write some mailets in Scala and demonstrate how it's done and then used in a running server.
+
+It also means writing and/or refactoring some server components in Scala, starting where it's the most relevant.
+
+### Positive Consequences 
+
+* Modernize parts of James code
+* Leverage Scala richer FP ecosystem and language to overcome Java limitations on that topic
+* Should attract people that would not like Java
+
+### Negative Consequences 
+
+* Adds even more knowledge requirements to contribute to James
+* Scala build time is longer than Java build time
+
+## Pros and Cons of the Options
+
+### Option 1: Let the users figure out how to make polyglot setups
+
+Pros:
+* We don't have anything new to do
+
+Cons:
+* It's boring, we like new challenges
+* Java is declining despite language modernization, it means in the long term, less and less people will contribute to James
```

Comment:
reference?
and if it's true then reference about Scala ascent?

Reply:
Some statistical references can strengthen the argument.

Reply:
Comparing https://insights.stackoverflow.com/survey/2018 to https://insights.stackoverflow.com/survey/2019 shows that Java popularity fall from 45.3% to 41.1%.
Github https://octoverse.github.com/#top-languages shows that Python is now more popular than Java.
TIOBE index ( https://www.tiobe.com/tiobe-index/ ) is -0.01% (no a big decline, for sure)

Scala is not growing neither but adding a new language creates new opportunity for contributions. BTW, Kotlin, which is growing quite fast is doing so because it's endorsed by Google for Android and yet it doesn't have stellar adoption.

BTW, and maybe it's missing from this ADR, one choose a language based on how it fits the problems needed to be solved. In this case, and I'm definitely biased for being more competent in Scala than in Kotlin, Kotlin would be a great language for James but Scala still looks superior given the maturity of the language and its ecosystem.